flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Hybrid source reader that delegates to the actual source reader.
+ *
+ * <p>This reader is setup with a sequence of underlying source readers. At a given point in time,
+ * one of these readers is active. Underlying readers are opened and closed on demand as determined
+ * by the enumerator, which selects the active reader via {@link SwitchSourceEvent}.
+ *
+ * <p>When the underlying reader has consumed all input, {@link HybridSourceReader} sends {@link
+ * SourceReaderFinishedEvent} to the coordinator and waits for the {@link SwitchSourceEvent}.
+ */
+public class HybridSourceReader<T> implements SourceReader<T, HybridSourceSplit> {
+    private static final Logger LOG = LoggerFactory.getLogger(HybridSourceReader.class);
+    private final SourceReaderContext readerContext;
+    private final List<SourceReader<T, ? extends SourceSplit>> chainedReaders;
+    private int currentSourceIndex = -1;
+    private SourceReader<T, ? extends SourceSplit> currentReader;
+    private CompletableFuture<Void> availabilityFuture;
+
+    public HybridSourceReader(
+            SourceReaderContext readerContext,
+            List<SourceReader<T, ? extends SourceSplit>> readers) {
+        this.readerContext = readerContext;
+        this.chainedReaders = readers;
+    }
+
+    @Override
+    public void start() {
+        setCurrentReader(0);
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput output) throws Exception {
+        InputStatus status = currentReader.pollNext(output);
+        if (status == InputStatus.END_OF_INPUT) {
+            // trap END_OF_INPUT if this wasn't the final reader
+            LOG.info(
+                    "End of input subtask={} sourceIndex={} {}",
+                    readerContext.getIndexOfSubtask(),
+                    currentSourceIndex,
+                    currentReader);
+            if (currentSourceIndex + 1 < chainedReaders.size()) {
+                // Signal the coordinator that the current reader has consumed all input and the
+                // next source can potentially be activated (after all readers are ready).
+                readerContext.sendSourceEventToCoordinator(
+                        new SourceReaderFinishedEvent(currentSourceIndex));
+                // More data will be available from the next reader.
+                // InputStatus.NOTHING_AVAILABLE requires us to complete the availability
+                // future after source switch to resume poll.
+                return InputStatus.NOTHING_AVAILABLE;
+            }
+        }
+        return status;
+    }
+
+    @Override
+    public List<HybridSourceSplit> snapshotState(long checkpointId) {
+        List<? extends SourceSplit> state = currentReader.snapshotState(checkpointId);
```

Comment:
I saw later that we snapshot a reader on switching, so it's probably fine as is.

Reply:
We are currently not returning to the last snapshot of a previous reader when splits are returned and we switch back to that reader. This is something that needs to be looked into further.

Reply:
@AHeise PTAL

It appears that this issue can be addressed by letting the reader drive the switching based on received splits: https://github.com/apache/flink/pull/15924/commits/c017f9b0b5da8950874bf20a33e2c3adf58daec2

This will also simplify the reset logic in the enumerator, as returned splits can now simply be assigned back to the reader. 

I still need to work on retaining the assigned splits when closing the underlying readers and include them into the snapshot.